Opentelemetry Sampling Rate Test (#710)
kyma-project/community: concepts/observability-strategy/configurable-tracing/pocs/tracing-sampling-test.md

Fig. 6 below show istio proxy resource consumption, during the test execution time, transferred data amount increased from around 10KB/s to 180KB/s in the peak time.

In the peak time, memory consumption increased from 61MB to 67 MB, CPU consumption increased from 0.100 to around 0.140.

Compare to the test **Scenario 2** amount of data transferred is almost 50% decreased, which belong to no tracing data could be transferred to the tracing collectors.

| ![Istio resource consumption](../assets/istio-100per-broken-svc-resource.jpg) |
| :--: |
| Fig. 6 Istio Proxy Resource Consumption |

Fig. 7 clearly show increased DNS requests and DNS lookup response time, to compared with **Scenario 2**
DNS request and DNS lookup response times are more than doubled.

Same behavior observable on the DNS cache metrics side, cache misses almost doubled to compared with **Scenario 2**.

| ![Kubernetes CoreDNS Overview](../assets/istio-100per-broken-svc-coredns-overview.jpg) |
| :--: |
| Fig. 7 Kubernetes CoreDNS Overview |

| ![Kubernetes CoreDNS Cache](../assets/istio-100per-broken-svc-coredns-cache.jpg) |
| :--: |
| Fig. 8 Kubernetes CoreDNS Cache |

#### Summary

> To compare with **Scenario 2**, overall network latency is bad, response times increased from below 1 second up to 9 second.

Reviewer: ```suggestion
Compared to **Scenario 2**, the overall network latency is bad, response times increased from below 1 second up to 9 seconds.
```

> Same behavior observable by the CoreDNS service, DNS requests and DNS lookups times increased up to 100% compared with **Scenario 2**

Reviewer: ```suggestion
The same behavior is observable for the CoreDNS service: Compared to **Scenario 2**, DNS requests and DNS lookup times increased by up to 100%.
```

### Scenario 4

> Fig.1 shown call execution summary (from client perspective) of test cluster with istio sampling rate 100% setup.

Reviewer: ```suggestion
Fig. 1 shows the call execution summary (from client perspective) of the test cluster with Istio sampling rate set to **100%**.
```

> Success rate of call execution is 99% and average response time is around **1 seconds.**

Reviewer: ```suggestion
The success rate of call execution is 99%, and the average response time is around **1 second**.
```

| ![Istio service detail](../assets/istio-100per-no-svc-servicedetail.jpg) |
| :--: |
| Fig. 3 Istio Service Overview |

| ![Service network detail 1](../assets/istio-100per-no-svc-servicedetail-2.jpg) |
| :--: |
| Fig. 4 Service Network Details |

| ![Service network detail 2](../assets/istio-100per-no-svc-servicedetail-3.jpg) |
| :--: |
| Fig. 5 Service Network Details |

Fig. 6 below show istio proxy resource consumption during the test execution time, bytes transferred during entire test increased from around 10KB/s to 77KB/s in the peak time.

Memory consumption of istio proxy increased from 66MB to 67 MB in peak time, CPU consumption, before test started was 0.050 in the peak time the value increased to around 0.100.

By test start, memory consumption as well as CPU consumption increased rapidly but decreased in 5 minutes to the normal values.

Data amount transferred with in the mesh compared to previous scenarios **Scenario 2** and **Scenario 3** significantly decreased,
which to lead back missing tracing collector service calls.

| ![Istio resource consumption](../assets/istio-100per-no-svc-resource.jpg) |
| :--: |
| Fig. 6 Istio Proxy Resource Consumption |

> Like **Scenario 3** Kubernetes CoreDNS service side DNS request per second increased significantly, from 2 packet per second to 8 packet per second.

Reviewer: ```suggestion
Fig. 7 shows an overview of the Kubernetes CoreDNS service. 
Like in **Scenario 3**, the DNS requests per second increased significantly, from 2 packet/sec to 8 packet/sec.
```

> DNS cache misses also doubled in the test execution time, which also explain increased DSN lookup response time, this value increased from 1 second to 4 seconds and response times up to 500ms during test execution time.

Reviewer: ```suggestion
During test execution, DNS cache misses also doubled, which also explains the increased DNS lookup response time. DNS lookup response time increased from 1 second to 4 seconds, and response times went up to 500ms.
```

| ![Kubernetes CoreDNS Overview](../assets/istio-100per-no-svc-coredns-overview.jpg) |
| :--: |
| Fig. 7 Kubernetes CoreDNS Overview |

| ![Kubernetes CoreDNS Cache](../assets/istio-100per-no-svc-coredns-cache.jpg) |
| :--: |
| Fig. 8 Kubernetes CoreDNS Cache |

#### Summary

CoreDNS service lookup times and DNS requests are increased significantly, overall network latency increased. 

### Conclusion

Table below compare all relevant metrics for all four scenarios.

> 

Reviewer: I ❤️ this overview!
I'll copy it into the introductory "Scenarios" section.

- **Scenario 1** with 1% sampling rate
- **Scenario 2** with 100% sampling rate
- **Scenario 3** with 100% sampling rate, collector service without end point
- **Scenario 4** with 100% sampling rate, no collector deployed

||Scenario 1|Scenario 2|Scenario 3|Scenario 4|
|:--|:--:|:--:|:--:|:--:|
|Success Rate|100%|99%|33%|99%|
|Global Request (Mesh)|20.2 ops/s|65.8 ops/s|72.8 ops/s|21.1 ops/s|
|Request per Service|6.20 ops/s|6.70 ops/s|7.20 ops/s|7.33 ops/s|
|Latency|400ms|950ms|9 sec|1 sec|
|Bytes Transferred|60KB/s|300KB/s|170KB/s|70KB/s|
|DNS Requests|3 p/s|3 p/s|8 p/s|8 p/s|
|DNS Lookups|150ms|150ms|500ms|700ms|

A 100% sampling rate will put more pressure on overall network communication, which will result to high latency and high load on collector services.

> Deployment without collectors (or broken collector service), will result CoreDNS service pressure, DNS request and DNS lookups will increase up to 400% compared to with collectors.

Reviewer: ```suggestion
Deployment without collectors (or a broken collector service) causes pressure on the Kubernetes CoreDNS service; DNS requests and DNS lookups increase up to 400% compared to scenarios with working collectors.
```

## Setup
### Opentelemetry
OpenTelemetry version 0.22.1 deployed on a Kyma cluster (version 2.6.0) with a minimal setup,
minimal setup mean :
- Standard deployment of OpenTelemetry Helm Chart version 0.22.1
- As receivers standard Jaeger, Zipkin, and OTLP configured
- As processor only memory limiter processor with standard configuration
- As exporter only log/info and/or NOP exporter to keep impact of OpenTelemetry as low as possible
- No andy additional extensions

### Sampling App Deployment

As sampling app, 3 serverless function deployed with istio sidecar injected and which calling each other in a chain like
**Extern Call -> FunctionA -> FunctionB -> FunctionC** to simulate a trace fully sampled with an external call.

All functions are deployed with NodeJs version 16.

### Call Simulator
As call simulator Gatling version 3.8.3 used.

## Scenario
All scenarios are having same OpenTelemetry deployment and configuration described above.

Same Sampling app used in all scenarios.

Same Gatling call simulator used in all scenarios, Gatling shall call **FunctionA** from extern with at least 5 simultaneous users up to 10 maximum.

> Call is simple URL call of FunctionA with no additional data or http headers to keep any influence of those on trace self.

Reviewer: ```suggestion
```
Added this to the previous suggestion.

> Call simulation should run 100 minutes long to put enough load on call chain and generate enough metrics to get precise result.

Reviewer: ```suggestion
```
Added this to the previous suggestion.

### Scenario 1

> Kyma standard (version 2.6.0) deployment with istio sampling rate configured to **1%** sampling, to observe istio behavior like resource consumption and throughput.

Reviewer: ```suggestion
Kyma standard deployment (version 2.6.0) with Istio sampling rate configured to **1%** sampling, to observe Istio behavior like resource consumption and throughput.
```

### Scenario 2 

> Kyma standard deployment from main branch with istio sampling rate configuration changed to **100%** sampling, to observe istio behavior like resource consumption and throughput.

Reviewer: ```suggestion
Kyma standard deployment from main branch with Istio sampling rate configuration changed to **100%** sampling, to observe Istio behavior like resource consumption and throughput.
```

### Scenario 3

> Like **Scenario 2** described above with additional configuration on **Jaeger** and **Zipkin** receivers services without endpoint.

Reviewer: ```suggestion
Like **Scenario 2** with additional configuration on **Jaeger** and **Zipkin** receiver services without valid endpoint. This scenario should investigates additional impact on Kubernetes components.
```

> Services **Jaeger** and **Zipkin** in this scenario are exist but pointing no valid endpoint. This scenario should focus additionally impact on kubernetes components.

Reviewer: ```suggestion
```
Isn't this sentence redundant to the one in line 44? "Jaeger and Zipkin exist but have no valid endpoint" is the message in both of them.

### Scenario 4

> Like Scenario 2 described above, this setup should run without collectors **Jaeger** and **Zipkin** to analyse istio behaviour and impact on kubernetes components like **CoreDNS**.

Reviewer: ```suggestion
Like Scenario 2, runs without collectors **Jaeger** and **Zipkin** to analyse Istio behaviour and impact on Kubernetes components like **CoreDNS**.
```

Scenario 2 is already without Jaeger and Zipkin, right? (I'm guessing this because Scenario 3 mentions specifically Jaeger and Zipkin as addition to the normal Scenario 2 setup). So how is Scenario 4 different from Scenario 2?

Author: Scenario 1 and Scenario 2 are having working Jaeger service, in the Scenario 3 Jaeger broken and Scenario 4 has no Jaeger or Zipkin deployed

## Result

> Overall analysis focus on istio behavior and resource consumption under different configuration and setup, in **Scenario 3** and **Scenario 4** additionally focus on kubernetes components.

Reviewer: Why is this sentence here and not after line 27? This sentence summarizes the scenarios, not the results.
Please consider moving this sentence between after line 27.

```suggestion
The overall analysis focuses on Istio behavior and resource consumption under different setups. In **Scenario 3** and **Scenario 4**, there's additional focus on Kubernetes components.
```
"configuration and setup" is redundant, one of both words is enough.

Author: I think setup is enough

Reviewer: I agree. But my question was this: 

> Why is this sentence here and not after line 27? This sentence summarizes the scenarios, not the results.
> Please consider moving this sentence between after line 27.

But never mind, I decided to simply copy the overview from the "Conclusion" section into the "Scenarios" section, which should clarify things.
